File: src/main/kotlin/rainbow/inner/function/mathfunction/simple/SimpleMathFunction.java
package rainbow.inner.function.mathfunction.simple;

import rainbow.inner.function.mathfunction.MathFunction;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.function.DoubleUnaryOperator;

/**
 * 简化版的MathFunction，对于任意一个维度进行线性处理。默认为第一个。
 *
 * @author Rainbow Yang
 */
public class SimpleMathFunction extends MathFunction {

    public SimpleMathFunction() {
    }

    public SimpleMathFunction(DoubleUnaryOperator... functions) {
        setFunction(0, functions);
    }

    public SimpleMathFunction(List<DoubleUnaryOperator> functions) {
        setFunction(0, functions);
    }

    public SimpleMathFunction(int index, DoubleUnaryOperator... functions) {
        setFunction(index, functions);
    }

    public SimpleMathFunction(int index, List<DoubleUnaryOperator> functions) {
        setFunction(index, functions);
    }

    public void setFunction(DoubleUnaryOperator... functions) {
        setFunction(0, functions);
    }

    public void setFunction(List<DoubleUnaryOperator> functions) {
        setFunction(0, functions);
    }

    public void setFunction(int index, DoubleUnaryOperator... functions) {
        setFunction(index, Arrays.asList(functions));
    }

    public void setFunction(int index, List<DoubleUnaryOperator> functions) {
        ArrayList<DoubleUnaryOperator> newFunctions = new ArrayList<>();
        newFunctions.addAll(functions);
        newFunctions.add(index, SELF);
        //todo
        addFunctions(newFunctions);
    }
}

File: src/main/kotlin/rainbow/function/mathfunction/simple/SimpleMathFunction.kt
package rainbow.function.mathfunction.simple

import rainbow.coordinate.function.mathfunction.DoubleUnaryOperator
import rainbow.function.mathfunction.MathFunction
import java.util.*

/**
 * 简化版的MathFunction，对于任意一个维度进行线性处理。默认为第一个。
 * @author Rainbow Yang
 */
abstract class SimpleMathFunction : MathFunction() {

    fun setFunction(function: DoubleUnaryOperator) = setFunction(0, listOf(function))

    fun setFunction(index: Int = 0, function: DoubleUnaryOperator) = setFunction(index, listOf(function))

    fun setFunction(index: Int = 0, vararg functions: DoubleUnaryOperator) = setFunction(index, functions.asList())

    fun setFunction(index: Int = 0, functions: List<DoubleUnaryOperator>) {
        val newFunctions = LinkedList<DoubleUnaryOperator>()
        newFunctions.addAll(functions)
        newFunctions.add(index, SELF)

        addFunctions(newFunctions)
    }
}
